processing/src/main/java/io/druid/segment/incremental/OffheapOakIncrementalIndex.java

```diff
@@ -0,0 +1,777 @@
+ */
+
+package io.druid.segment.incremental;
+
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import io.druid.data.input.InputRow;
+import io.druid.data.input.MapBasedRow;
+import io.druid.data.input.Row;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.logger.Logger;
+import io.druid.java.util.common.parsers.ParseException;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.query.aggregation.PostAggregator;
+import io.druid.segment.column.ColumnCapabilitiesImpl;
+import io.druid.segment.column.ValueType;
+import io.druid.segment.ColumnValueSelector;
+import io.druid.segment.DimensionHandler;
+import io.druid.segment.DimensionIndexer;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.List;
+import java.util.Comparator;
+import java.util.function.Function;
+import java.util.concurrent.atomic.AtomicInteger;
+import oak.OakMapOffHeapImpl;
+import oak.OakMap;
+import oak.CloseableIterator;
+
+import javax.annotation.Nullable;
+
+
+/**
+ */
+public class OffheapOakIncrementalIndex extends
+        io.druid.segment.incremental.InternalDataIncrementalIndex<BufferAggregator>
+{
+
+  private static final Logger log = new Logger(OffheapOakIncrementalIndex.class);
+
+  // When serializing an object from IncrementalIndexRow.dims, we use:
+  // 1. 4 bytes for representing its type (Double, Float, Long or String)
+  // 2. 8 bytes for saving its value or the array position and length (in the case of String)
+  static final Integer ALLOC_PER_DIM = 12;
+  static final Integer NO_DIM = -1;
+  static final Integer TIME_STAMP_INDEX = 0;
+  static final Integer DIMS_LENGTH_INDEX = TIME_STAMP_INDEX + Long.BYTES;
+  static final Integer DIMS_INDEX = DIMS_LENGTH_INDEX + Integer.BYTES;
+
+  OakMapOffHeapImpl oak;
+  private final int maxRowCount;
+  private String outOfRowsReason = null;
+  private OffheapAggsManager aggsManager;
+
+  OffheapOakIncrementalIndex(
+          io.druid.segment.incremental.IncrementalIndexSchema incrementalIndexSchema,
+          boolean deserializeComplexMetrics,
+          boolean reportParseExceptions,
+          boolean concurrentEventAdd,
+          int maxRowCount,
+          int chunkMaxItems,
+          int chunkBytesPerItem
+  )
+  {
+    super(incrementalIndexSchema, deserializeComplexMetrics, reportParseExceptions,
+            concurrentEventAdd);
+
+    IncrementalIndexRow minIncrementalIndexRow = getMinIncrementalIndexRow();
+    oak = new OakMapOffHeapImpl(new IncrementalIndexRowByteBuffersComp(dimensionDescsList),
+            minIncrementalIndexRow,
+            new OffheapOakCreateKeyConsumer(dimensionDescsList),
+            new OffheapOakKeyCapacityCalculator(dimensionDescsList),
+            chunkMaxItems,
+            chunkBytesPerItem);
```

Comment:
Not so related to Druid, but why the size of the chunk need to be provided here? This need to be configurable, but doesn't look reasonable to be part of constructor... ?

Reply:
Actually both chunkMaxItems and chunkBytesPerItem looks to me weird to be provided in constructor. It should be eliminated from the Oak Constructor and we should give user other ways to influence the defauklt (if needed).